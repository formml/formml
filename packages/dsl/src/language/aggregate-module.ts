import { EmptyFileSystem, inject } from 'langium'
import {
  DefaultSharedModuleContext,
  LangiumSharedServices,
  createDefaultModule,
  createDefaultSharedModule,
} from 'langium/lsp'

import { FormMLModule, FormMLServices } from './formml/formml-module.js'
import { registerFormMLValidationChecks } from './formml/formml-validator.js'
import {
  FormMLDeclarationModule,
  FormMLDeclarationServices,
} from './formml-declaration/formml-declaration-module.js'
import {
  FormMLDeclarationGeneratedModule,
  FormMLGeneratedModule,
  FormMLGeneratedSharedModule,
} from './generated/module.js'
import LongestMatchServiceRegistry from './overrides/LongestMatchServiceRegistry.js'

/**
 * Create the full set of services required by Langium.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection
 * @returns An object wrapping the shared services and the language-specific services
 */
export function createAggregateServices(context: DefaultSharedModuleContext): {
  FormML: FormMLServices
  FormMLDeclaration: FormMLDeclarationServices
  shared: LangiumSharedServices
} {
  const shared = inject(
    createDefaultSharedModule(context),
    FormMLGeneratedSharedModule,
    {
      ServiceRegistry: () => new LongestMatchServiceRegistry(),
    },
  )
  const FormML = inject(
    createDefaultModule({ shared }),
    FormMLGeneratedModule,
    FormMLModule,
  )
  const FormMLDeclaration = inject(
    createDefaultModule({ shared }),
    FormMLDeclarationGeneratedModule,
    FormMLDeclarationModule,
  )
  shared.ServiceRegistry.register(FormML)
  shared.ServiceRegistry.register(FormMLDeclaration)

  registerFormMLValidationChecks(FormML)
  // registerFormMLDeclarationValidationChecks(FormMLDeclaration)

  if (!context.connection) {
    // We don't run inside a language server
    // Therefore, initialize the configuration provider instantly
    void shared.workspace.ConfigurationProvider.initialized({})
  }
  return { FormML, FormMLDeclaration, shared }
}

export const createInMemoryAggregateServices = () =>
  createAggregateServices(EmptyFileSystem)
