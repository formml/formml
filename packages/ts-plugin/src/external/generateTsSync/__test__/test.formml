form testForm {
  num  numField
  text textField
}
