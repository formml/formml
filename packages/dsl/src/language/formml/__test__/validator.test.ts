import { clearDocuments, parseHelper } from 'langium/test'

import { createInMemoryAggregateServices } from '../../aggregate-module.js'
import { FormMLDeclaration, FormMLSchema } from '../../index.js'

describe('formml validator', () => {
  const services = createInMemoryAggregateServices()

  afterEach(async () => {
    await clearDocuments(services.shared)
  })

  const loadDeclaration = (input: string, uri: string) =>
    parseHelper<FormMLDeclaration>(services.FormMLDeclaration)(input, {
      documentUri: uri,
    })
      .then((x) => x.parseResult)
      .then((r) =>
        r.lexerErrors.length > 0 || r.parserErrors.length > 0
          ? Promise.reject(
              new Error(
                'Parsing failed with errors:\n' +
                  [...r.lexerErrors, ...r.parserErrors]
                    .map((e) => e.message)
                    .join('\n'),
              ),
            )
          : r.value,
      )
  const parser = (input: string) =>
    parseHelper<FormMLSchema>(services.FormML)(input, { validation: true })

  describe('annotation', () => {
    beforeEach(async () => {
      const annotations = `
        annot fun required()
        annot fun range(min, max)
        annot fun any(name1, name2, name3, name4, name5, name6, name7, name8, name9, name10)
      `
      await loadDeclaration(annotations, 'file:///builtin-annotations.d.formml')
    })

    test('should no error when annotation is valid', async () => {
      const input = `
        form ExampleForm {
          @range(0, max: 100)
          num numberField
        }
      `
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(0)
    })

    test('should error when called annotation is not defined', async () => {
      const input = `
        form ExampleForm {
          @unknown
          num numberField
        }
      `
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(1)
      expect(diagnostics).toMatchSnapshot()
    })

    test('should error when annotation call is not immediately followed by @ sign', async () => {
      const input = `
        form ExampleForm {
          @     required
          num numberField
        }
      `
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(1)
      expect(diagnostics).toMatchSnapshot()
    })

    test('should error when named arguments appear before positional arguments', async () => {
      const input = `
        form ExampleForm {
          @range(min: 0, 100)
          num numberField
        }
      `
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(1)
      expect(diagnostics).toMatchSnapshot()
    })

    test('should error when named arguments appear before positional arguments - multiple appearances', async () => {
      const input = `
        form ExampleForm {
          @any(1, name2: "value2", 3, name4: "value4", name5: "value5", 6, name7: "value7")
          num numberField
        }
      `
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(3)
      expect(diagnostics).toMatchSnapshot()
    })

    test('should no error given insufficient arguments', async () => {
      const input = `
        form ExampleForm {
          @range(0)
          num numberField
        }
      `
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(0)
    })

    test('should error given too many arguments', async () => {
      const input = `
        form ExampleForm {
          @range(0, 100, 200)
          num numberField
        }
      `
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(1)
      expect(diagnostics).toMatchSnapshot()
    })

    test('should error given unknown named argument', async () => {
      const input = `
        form ExampleForm {
          @range(0, unknown: 100)
          num numberField
        }
      `
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(1)
      expect(diagnostics).toMatchSnapshot()
    })

    test('should error when re-assigning to same parameter', async () => {
      const input = `
        form ExampleForm {
          @range(min: 10, min: 100)
          num numberField
        }
      `
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(1)
      expect(diagnostics).toMatchSnapshot()
    })

    test('should error when re-assigning to same parameter - mix of named and positional arguments', async () => {
      const input = `
        form ExampleForm {
          @range(10, min: 100)
          num numberField
        }
      `
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(1)
      expect(diagnostics).toMatchSnapshot()
    })
  })
})
