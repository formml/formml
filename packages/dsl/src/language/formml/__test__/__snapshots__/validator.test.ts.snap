// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formml validator > annotation > should error given too many arguments 1`] = `
[
  {
    "code": undefined,
    "codeDescription": undefined,
    "data": undefined,
    "message": "Expected 2 arguments, but got 3.",
    "range": {
      "end": {
        "character": 28,
        "line": 2,
      },
      "start": {
        "character": 25,
        "line": 2,
      },
    },
    "relatedInformation": undefined,
    "severity": 1,
    "source": "formml",
    "tags": undefined,
  },
]
`;

exports[`formml validator > annotation > should error given unknown named argument 1`] = `
[
  {
    "code": undefined,
    "codeDescription": undefined,
    "data": undefined,
    "message": "Unknown parameter "unknown", expected one of "min" | "max".",
    "range": {
      "end": {
        "character": 32,
        "line": 2,
      },
      "start": {
        "character": 20,
        "line": 2,
      },
    },
    "relatedInformation": undefined,
    "severity": 1,
    "source": "formml",
    "tags": undefined,
  },
]
`;

exports[`formml validator > annotation > should error when annotation call is not immediately followed by @ sign 1`] = `
[
  {
    "code": undefined,
    "codeDescription": undefined,
    "data": undefined,
    "message": "Expect an annotation call immediately after "@" sign, but found whitespaces.",
    "range": {
      "end": {
        "character": 16,
        "line": 2,
      },
      "start": {
        "character": 11,
        "line": 2,
      },
    },
    "relatedInformation": undefined,
    "severity": 1,
    "source": "formml",
    "tags": undefined,
  },
]
`;

exports[`formml validator > annotation > should error when called annotation is not defined 1`] = `
[
  {
    "code": undefined,
    "codeDescription": undefined,
    "data": {
      "code": "linking-error",
      "containerType": "Annotation",
      "property": "call",
      "refText": "unknown",
    },
    "message": "Could not resolve reference to AnnotationDeclaration named 'unknown'.",
    "range": {
      "end": {
        "character": 18,
        "line": 2,
      },
      "start": {
        "character": 11,
        "line": 2,
      },
    },
    "relatedInformation": undefined,
    "severity": 1,
    "source": "formml",
    "tags": undefined,
  },
]
`;

exports[`formml validator > annotation > should error when named arguments appear before positional arguments - multiple appearances 1`] = `
[
  {
    "code": undefined,
    "codeDescription": undefined,
    "data": undefined,
    "message": "Named argument can only appear after all positional arguments.",
    "range": {
      "end": {
        "character": 33,
        "line": 2,
      },
      "start": {
        "character": 18,
        "line": 2,
      },
    },
    "relatedInformation": undefined,
    "severity": 1,
    "source": "formml",
    "tags": undefined,
  },
  {
    "code": undefined,
    "codeDescription": undefined,
    "data": undefined,
    "message": "Named argument can only appear after all positional arguments.",
    "range": {
      "end": {
        "character": 53,
        "line": 2,
      },
      "start": {
        "character": 38,
        "line": 2,
      },
    },
    "relatedInformation": undefined,
    "severity": 1,
    "source": "formml",
    "tags": undefined,
  },
  {
    "code": undefined,
    "codeDescription": undefined,
    "data": undefined,
    "message": "Named argument can only appear after all positional arguments.",
    "range": {
      "end": {
        "character": 70,
        "line": 2,
      },
      "start": {
        "character": 55,
        "line": 2,
      },
    },
    "relatedInformation": undefined,
    "severity": 1,
    "source": "formml",
    "tags": undefined,
  },
]
`;

exports[`formml validator > annotation > should error when named arguments appear before positional arguments 1`] = `
[
  {
    "code": undefined,
    "codeDescription": undefined,
    "data": undefined,
    "message": "Named argument can only appear after all positional arguments.",
    "range": {
      "end": {
        "character": 23,
        "line": 2,
      },
      "start": {
        "character": 17,
        "line": 2,
      },
    },
    "relatedInformation": undefined,
    "severity": 1,
    "source": "formml",
    "tags": undefined,
  },
]
`;

exports[`formml validator > annotation > should error when re-assigning to same parameter 1`] = `
[
  {
    "code": undefined,
    "codeDescription": undefined,
    "data": undefined,
    "message": "Duplicate assignment to parameter "min".",
    "range": {
      "end": {
        "character": 34,
        "line": 2,
      },
      "start": {
        "character": 26,
        "line": 2,
      },
    },
    "relatedInformation": undefined,
    "severity": 1,
    "source": "formml",
    "tags": undefined,
  },
]
`;
